### PR TITLE
Kafka reactive getting started

### PR DESCRIPTION
**Pull request by ozangunalp**
**Check list**:

Your pull request:

- [x] targets the `development` branch
- [x] uses the `999-SNAPSHOT` version of Quarkus
- [x] has tests (`mvn clean test`)
- [x] works in native (`mvn clean package -Pnative`)
- [x] has native tests (`mvn clean verify -Pnative`)
- [ ] makes sure the associated guide must not be updated
- [x] links the guide update pull request (if needed) see https://github.com/quarkusio/quarkus/pull/17830
- [x] updates or creates the `README.md` file (with build and run instructions)
- [ ] for new quickstart, is located in the directory _component-quickstart_
- [ ] for new quickstart, is added to the root `pom.xml` and `README.md`

### CODE REVIEW

**Comment by Ladicek** [comment]
Looks OK, just wondering why you remove the `PriceConverter`? It doesn't make much sense for sure, but it shows the 3rd important kind of Reactive Messaging component:

- producer (`@Outgoing`)
- consumer (`@Incoming`)
- processor (`@Incoming @Outgoing`)

**Comment by ozangunalp** [comment]
You are right. I wanted to simplify things but went overboard. The new example doesn't show the `@Incoming` at all. 
I'll continue to work for a better example.

File: kafka-quickstart/producer/src/main/resources/application.properties [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Configure the Kafka sink (we write to it)
+mp.messaging.outgoing.quote-requests.connector=smallrye-kafka
+mp.messaging.outgoing.quote-requests.topic=quote-requests
+mp.messaging.outgoing.quote-requests.value.serializer=org.apache.kafka.common.serialization.StringSerializer
+
+# Configure the Kafka source (we read from it)
+mp.messaging.incoming.quotes.connector=smallrye-kafka
+mp.messaging.incoming.quotes.topic=quotes
```
Comment (cescoffier) [inline kafka-quickstart/producer/src/main/resources/application.properties:8]: not required.

File: kafka-quickstart/producer/src/main/resources/application.properties [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+mp.messaging.incoming.quotes.value.deserializer=org.acme.kafka.model.QuoteDeserializer
+mp.messaging.incoming.quotes.health-readiness-enabled=false
```
Comment (cescoffier) [inline kafka-quickstart/producer/src/main/resources/application.properties:10]: can be removed.

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: kafka-quickstart/processor/src/test/java/org/acme/kafka/processor/QuoteProcessorTest.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+public class QuoteProcessorTest {
+
+    @Test
+    void name() {
+
```
[the remark this answers is not in the document]
Comment (ozangunalp) [reply kafka-quickstart/processor/src/test/java/org/acme/kafka/processor/QuoteProcessorTest.java:12]: I forgot to add a failing test to not forget adding a proper test :)

File: kafka-quickstart/producer/src/main/resources/application.properties [hunk 4]
```diff
@@ -0,0 +1,10 @@
+# Configure the Kafka sink (we write to it)
+mp.messaging.outgoing.quote-requests.connector=smallrye-kafka
+mp.messaging.outgoing.quote-requests.topic=quote-requests
+mp.messaging.outgoing.quote-requests.value.serializer=org.apache.kafka.common.serialization.StringSerializer
```
[the remark this answers is not in the document]
Comment (ozangunalp) [reply kafka-quickstart/producer/src/main/resources/application.properties:4]: This seems to be the only one that is not detected. I need to dig more on this.

File: kafka-quickstart/producer/src/main/resources/application.properties [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+
+# Configure the Kafka source (we read from it)
+mp.messaging.incoming.quotes.connector=smallrye-kafka
+mp.messaging.incoming.quotes.topic=quotes
+mp.messaging.incoming.quotes.value.deserializer=org.acme.kafka.model.QuoteDeserializer
```
[the remark this answers is not in the document]
Comment (ozangunalp) [reply kafka-quickstart/producer/src/main/resources/application.properties:9]: I went with including serde configs. I'll remove all that is autodetected.

File: kafka-quickstart/processor/pom.xml [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+                                    </systemPropertyVariables>
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+        <profile>
+            <id>docker</id>
```
[the remark this answers is not in the document]
Comment (ozangunalp) [reply kafka-quickstart/processor/pom.xml:117]: Not used, it was there before. I'll remove it

File: kafka-quickstart/README.md [hunk 7]
```diff
@@ -38,4 +65,5 @@ As you are running in _prod_ mode, you need a Kafka cluster. You can follow the
 
 Then run with:
```
[the remark this answers is not in the document]
Comment (ozangunalp) [reply kafka-quickstart/README.md:66]: It is a bit implicit isn't it?

File: kafka-quickstart/README.md [hunk 8]
```diff
@@ -3,30 +3,57 @@ Quarkus Kafka Quickstart
...
+* `QuoteProcessor` consumes quote request id's from the `quote-requests` Kafka topic and responds back to the `quotes` topic with a `Quote` object containing a random price.
 
-* `PriceGenerator` - a bean generating random price. They are sent to a Kafka topic
-* `PriceConverter` - on the consuming side, the `PriceConverter` receives the Kafka message and convert the price.
-The result is sent to an in-memory stream of data
-* `PriceResource`  - the `PriceResource` retrieves the in-memory stream of data in which the converted prices are sent and send these prices to the browser using Server-Sent Events.
+Interactions with Kafka is managed by MicroProfile Reactive Messaging.
+Application configurations are located in `application.properties` files.
+
+## Running the application in Docker
```
[the remark this answers is not in the document]
Comment (ozangunalp) [reply kafka-quickstart/README.md:42]: We ran the dev mode in the beginning

File: kafka-quickstart/README.md [hunk 9]
```diff
@@ -3,39 +3,86 @@ Quarkus Kafka Quickstart
 
 This project illustrates how you can interact with Apache Kafka using MicroProfile Reactive Messaging.
```
Comment (cescoffier) [inline kafka-quickstart/README.md:4]:
```suggestion
This project illustrates how Quarkus application you can interact with Apache Kafka using Reactive Messaging.
```

File: kafka-quickstart/README.md [hunk 10]
```diff
@@ -3,39 +3,86 @@ Quarkus Kafka Quickstart
...
+The application is composed of two applications communicating through Kafka.
+Interactions with Kafka is managed by MicroProfile Reactive Messaging.
+
+They can be started in dev mode using:
+
+```bash
+mvn -f producer quarkus:dev
+```
+
+and
```
Comment (cescoffier) [inline kafka-quickstart/README.md:17]:
```suggestion
and, in another terminal:
```

File: kafka-quickstart/README.md [hunk 11]
```diff
@@ -3,39 +3,86 @@ Quarkus Kafka Quickstart
...
+You can send quote requests and observe received quotes.
 
 ## Anatomy
 
-In addition to the `prices.html` page, the application is composed by 3 components:
+The application is composed of the following components:
+
+#### Producer
+
+The _producer_ application sends data to the Kafka broker.
```
Comment (cescoffier) [inline kafka-quickstart/README.md:34]:
```suggestion
The _producer_ application receives requests from the user (from HTTP) and sends data to the Kafka broker.
```

File: kafka-quickstart/README.md [hunk 12]
```diff
@@ -3,39 +3,86 @@ Quarkus Kafka Quickstart
...
-* `PriceResource`  - the `PriceResource` retrieves the in-memory stream of data in which the converted prices are sent and send these prices to the browser using Server-Sent Events.
+* `QuoteProducer` generates uniquely identified quote requests and sends them to the Kafka topic `quote-requests`.
+It also consumes the Kafka topic `quotes` and relays received messages to the browser using Server-Sent Events.
+* `quotes.html` sends quote requests to the previous endpoint and updates quotes with received prices.
 
-The interaction with Kafka is managed by MicroProfile Reactive Messaging.
-The configuration is located in the application configuration.
+#### Processor
+
+The _processor_ application receives data from Kafka.
```
Comment (cescoffier) [inline kafka-quickstart/README.md:43]:
```suggestion
The _processor_ application receives data from Kafka, processes them, and writes the result into the `quotes` Kafka topic.
```

File: kafka-quickstart/processor/src/main/java/org/acme/kafka/processor/QuotesProcessor.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+import javax.enterprise.context.ApplicationScoped;
+
+import org.acme.kafka.model.Quote;
+import org.eclipse.microprofile.reactive.messaging.Incoming;
+import org.eclipse.microprofile.reactive.messaging.Outgoing;
+
+import io.smallrye.reactive.messaging.annotations.Blocking;
+
+/**
+ * A bean consuming data from the "request" Kafka topic and giving out a random quote.
```
Comment (cescoffier) [inline kafka-quickstart/processor/src/main/java/org/acme/kafka/processor/QuotesProcessor.java:14]: wasn't it the quote-requests topic?

Comment (cescoffier) [reply kafka-quickstart/processor/src/main/java/org/acme/kafka/processor/QuotesProcessor.java:14]: also mentioned that the topic is mapped to the "requests" channel.

File: kafka-quickstart/processor/src/main/resources/application.properties [hunk 14]
```diff
@@ -0,0 +1,10 @@
+%dev.quarkus.http.port=8081
```
Comment (cescoffier) [inline kafka-quickstart/processor/src/main/resources/application.properties:1]: Why is this needed? 8081 is the port used in tests, so should probably be 8082.

File: kafka-quickstart/processor/src/main/resources/application.properties [hunk 15]
```diff
@@ -0,0 +1,10 @@
...
+
+# Configure the incoming Kafka topic quote-requests (we read from it)
```
Comment (cescoffier) [inline kafka-quickstart/processor/src/main/resources/application.properties:3]:
```suggestion
# Configure the incoming `quote-requests` Kafka topic 
```

File: kafka-quickstart/processor/src/main/resources/application.properties [hunk 16]
```diff
@@ -0,0 +1,10 @@
...
+mp.messaging.incoming.requests.connector=smallrye-kafka
+mp.messaging.incoming.requests.topic=quote-requests
+mp.messaging.incoming.requests.auto.offset.reset=earliest
+
+# Configure the outgoing Kafka topic quotes (we write to it)
```
Comment (cescoffier) [inline kafka-quickstart/processor/src/main/resources/application.properties:8]:
```suggestion
# Configure the outgoing `quotes` Kafka topic
```

File: kafka-quickstart/processor/src/main/resources/application.properties [hunk 17]
```diff
@@ -0,0 +1,10 @@
...
+mp.messaging.outgoing.quotes.connector=smallrye-kafka
+mp.messaging.outgoing.quotes.value.serializer=io.quarkus.kafka.client.serialization.ObjectMapperSerializer
```
Comment (cescoffier) [inline kafka-quickstart/processor/src/main/resources/application.properties:10]: no discovery?

File: kafka-quickstart/processor/src/test/java/org/acme/kafka/processor/QuoteProcessorTest.java [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+    @Inject
+    @Identifier("default-kafka-broker")
+    Map<String, Object> kafkaConfig;
+
+    KafkaProducer<String, String> quoteRequestProducer;
+    KafkaConsumer<String, Quote> quoteConsumer;
+
+    @BeforeEach
+    void beforeAll() {
+        System.out.println(kafkaConfig);
```
Comment (cescoffier) [inline kafka-quickstart/processor/src/test/java/org/acme/kafka/processor/QuoteProcessorTest.java:40]: remove the trace.

File: kafka-quickstart/processor/src/test/java/org/acme/kafka/processor/QuoteProcessorTest.java [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+
+@QuarkusTest
+public class QuoteProcessorTest {
+
+    @Inject
+    @Identifier("default-kafka-broker")
+    Map<String, Object> kafkaConfig;
+
+    KafkaProducer<String, String> quoteRequestProducer;
+    KafkaConsumer<String, Quote> quoteConsumer;
```
Comment (cescoffier) [inline kafka-quickstart/processor/src/test/java/org/acme/kafka/processor/QuoteProcessorTest.java:36]: shouldn't you close these objects?

File: kafka-quickstart/processor/src/test/java/org/acme/kafka/processor/QuoteProcessorTest.java [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+
+    @BeforeEach
+    void beforeAll() {
+        System.out.println(kafkaConfig);
+        quoteConsumer = new KafkaConsumer<>(consumerConfig(), new StringDeserializer(), new ObjectMapperDeserializer<>(Quote.class));
+        quoteRequestProducer = new KafkaProducer<>(kafkaConfig, new StringSerializer(), new StringSerializer());
+    }
+
+    Properties consumerConfig() {
+        Properties properties = new Properties();
```
Comment (cescoffier) [inline kafka-quickstart/processor/src/test/java/org/acme/kafka/processor/QuoteProcessorTest.java:46]: unrelated to the review, but it's where I believe we can improve our testing support. We could "automagically" compute this configuration from the producer config.

Comment (cescoffier) [reply kafka-quickstart/processor/src/test/java/org/acme/kafka/processor/QuoteProcessorTest.java:46]: Alternatively, we could expose the managed Kafka consumer/producer/admin as bean.

File: kafka-quickstart/producer/src/main/java/org/acme/kafka/producer/QuotesResource.java [hunk 21]
```diff
@@ -0,0 +1,52 @@
...
+    @POST
+    @Path("/request")
+    @Produces(MediaType.TEXT_PLAIN)
+    public String createRequest() {
+        UUID uuid = UUID.randomUUID();
+        quoteRequestEmitter.send(uuid.toString());
+        return uuid.toString();
+    }
+
+    @Inject
```
Comment (cescoffier) [inline kafka-quickstart/producer/src/main/java/org/acme/kafka/producer/QuotesResource.java:40]: in this case, @Inject can be omitted (@Channel is a qualifier)

File: kafka-quickstart/producer/src/main/java/org/acme/kafka/producer/QuotesResource.java [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+import org.jboss.logging.Logger;
+
+import io.smallrye.mutiny.Multi;
+
+@Path("/quotes")
+public class QuotesResource {
+
+    private static final Logger LOG = Logger.getLogger(QuotesResource.class);
+
+    @Inject
```
Comment (cescoffier) [inline kafka-quickstart/producer/src/main/java/org/acme/kafka/producer/QuotesResource.java:24]: in this case, @Inject can be omitted (@Channel is a qualifier)

File: kafka-quickstart/producer/src/main/java/org/acme/kafka/producer/QuotesResource.java [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+    @Channel("quotes")
+    Multi<Quote> quotes;
+
+    /**
+     * Endpoint retrieving the "quotes" Kafka topic and sending the items to a server sent event.
+     */
+    @GET
+    @Produces(MediaType.SERVER_SENT_EVENTS) // denotes that server side events (SSE) will be produced
+    public Multi<Quote> stream() {
+        return quotes.onItem().invoke(quote -> LOG.infov("Received quote {0}", quote));
```
Comment (cescoffier) [inline kafka-quickstart/producer/src/main/java/org/acme/kafka/producer/QuotesResource.java:50]:
```suggestion
        return quotes.log();
```

So, you avoid having to have a logger defined in the class.

File: kafka-quickstart/producer/src/main/resources/application.properties [hunk 24]
```diff
@@ -0,0 +1,5 @@
+# Configure the outgoing Kafka topic quote-requests (we write to it)
```
Comment (cescoffier) [inline kafka-quickstart/producer/src/main/resources/application.properties:1]:
```suggestion
# Configure the outgoing `quote-requests` Kafka topic
```

File: kafka-quickstart/producer/src/main/resources/application.properties [hunk 25]
```diff
@@ -0,0 +1,5 @@
...
+mp.messaging.outgoing.quote-requests.connector=smallrye-kafka
+
+# Configure the incoming Kafka topic quotes (we read from it)
```
Comment (cescoffier) [inline kafka-quickstart/producer/src/main/resources/application.properties:4]:
```suggestion
# Configure the incoming `quotes` Kafka topic
```

File: kafka-quickstart/producer/src/test/java/org/acme/kafka/producer/QuotesResourceTest.java [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+    @Test
+    void testQuotesEventStream() {
+        String body = given()
+                .when()
+                .post("/quotes/request")
+                .then()
+                .statusCode(200)
+                .extract().body()
+                .asString();
+        assertDoesNotThrow(() -> UUID.fromString(body));
```
Comment (cescoffier) [inline kafka-quickstart/producer/src/test/java/org/acme/kafka/producer/QuotesResourceTest.java:24]: does this throw is the body is `null`?

**Review by cescoffier — CHANGES_REQUESTED** [review]

File: kafka-quickstart/producer/src/main/java/org/acme/kafka/producer/QuotesResource.java (hunk 23 above)
Comment (ozangunalp) [reply kafka-quickstart/producer/src/main/java/org/acme/kafka/producer/QuotesResource.java:50]: 🤯

File: kafka-quickstart/producer/src/test/java/org/acme/kafka/producer/QuotesResourceTest.java (hunk 26 above)
Comment (ozangunalp) [reply kafka-quickstart/producer/src/test/java/org/acme/kafka/producer/QuotesResourceTest.java:24]: Yes it'll throw an npe

**Review by cescoffier — APPROVED** [review]
